src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/DlsFlsValveImpl.java

```diff
@@ -140,4 +157,136 @@ public void handleSearchContext(SearchContext context, ThreadPool threadPool, Na
 
     }
 
+    @Override
+    public void onQueryPhase(SearchContext searchContext, long tookInNanos) {
+        QuerySearchResult queryResult = searchContext.queryResult();
+        if (queryResult == null) {
+            return;
+        }
+
+        DelayableWriteable<InternalAggregations> aggregationsDelayedWritable = queryResult.aggregations();
+        if (aggregationsDelayedWritable == null) {
+            return;
+        }
+
+        InternalAggregations aggregations = aggregationsDelayedWritable.expand();
+        if (aggregations == null) {
+            return;
+        }
+
+        if (areBucketKeysDistinct(aggregations)) {
+            return;
+        }
+
+        log.debug("Found buckets with equal keys. Merging these buckets: {}", aggregations);
+
+        // TODO check order
+
+        queryResult.aggregations(InternalAggregations.from(StreamSupport.stream(aggregations.spliterator(), false)
```

Comment:
Curious what is going on here, can you possibly elaborate why we are aggregated query results here?

Reply:
Starting with 7.10 ES assumes that buckets are strongly ordered by keys. Applying mask breaks strong ordering as some keys become the same. It is necessary to aggregate buckets with the same keys.